### PR TITLE
EVM-C v4

### PR DESCRIPTION
**Pull request by chfast**
EVMJIT part: https://github.com/ethereum/evmjit/pull/118

This includes hardcore vector created with placement new operator. This so more a proof of concept, we can reconsider replacing it. But this is also fun!

### CODE REVIEW

File: libevm/JitVM.cpp [hunk 1]
```diff
@@ -131,12 +132,13 @@ void getBlockHash(evm_uint256be* o_hash, evm_env* _envPtr, int64_t _number)
...
-int64_t call(
-	evm_env* _opaqueEnv,
-	evm_message const* _msg,
-	uint8_t* _outputData,
-	size_t _outputSize
-) noexcept
+void releaseVectorInPayload(evm_result const* _result)
+{
+	auto* b = reinterpret_cast<bytes const*>(&_result->payload.bytes);
+	b->~bytes();
```
Comment (chriseth) [inline libevm/JitVM.cpp]: This is very unusual and most of the time dangerous. If there is no other way than to call the destructor manually, please at least add a comment.

Comment (chfast) [inline libevm/JitVM.cpp]: I added a comments about it, and now the "destructor" code is very close to the "constructor" code (using non-capturing lambda).

File: libevm/JitVM.cpp [hunk 2]
```diff
@@ -173,15 +177,35 @@ int64_t call(
...
+	// when finished with it.
+
+	// First assign reference. References are not invalidated when vector
+	// of bytes is moved. See `.takeBytes()` below.
+	o_result->output_data = output.data();
+	o_result->output_size = output.size();
+
+	// Place a new vector of bytes containing output in result's reserved memory.
+	static_assert(sizeof(bytes) <= sizeof(o_result->reserved), "Vector is too big");
+	new(&o_result->reserved) bytes(output.takeBytes());
```
Comment (gumb0) [inline libevm/JitVM.cpp]: Does this placement new magic give us any significant advantage?
We could just save a pointer to dynamically allocated `bytes`, like `o_result->context = new bytes(output.takeBytes())`, then just `delete` it in release function.

Or even `o_result->context = new owning_bytes_ref(move(output))`, then we don't need `owning_bytes_ref::takeBytes`

Comment (chfast) [inline libevm/JitVM.cpp]: Yes, both options are good and safe.

The advantage of placement new is there is no dynamic memory allocation. We use some memory of evm_result object. I really wanted to try if it is possible (this affects evm_result design). But I'm happy to use a safer option at this point.

File: libevm/JitVM.cpp [hunk 3]
```diff
@@ -154,13 +148,23 @@ int64_t call(
...
 		std::tie(addr, std::ignore) = env.create(value, gas, input, {});
-		auto gasLeft = static_cast<int64_t>(gas);
+		o_result->gas_left = static_cast<int64_t>(gas);
+		o_result->release = nullptr;
 		if (addr)
-			std::copy(addr.begin(), addr.end(), _outputData);
+		{
+			o_result->code = EVM_SUCCESS;
+			// Use reserved data to store the address.
+			std::copy(addr.begin(), addr.end(), o_result->reserved.data);
```
Comment (gumb0) [inline libevm/JitVM.cpp]: There was an `assert` to check that `addr` fits into `_outputData`, which you deleted. Maybe add here `assert` or `static_assert` that `o_result->reserved.data` is at least 20 bytes

Comment (chfast) [inline libevm/JitVM.cpp]: Will be done. Good advice.

File: libevm/JitVM.cpp [hunk 4]
```diff
@@ -127,25 +127,19 @@ void getTxContext(evm_tx_context* result, evm_env* _opaqueEnv) noexcept
...
 	*o_hash = toEvmC(env.blockHash(_number));
 }
 
-int64_t call(
-	evm_env* _opaqueEnv,
-	evm_message const* _msg,
-	uint8_t* _outputData,
-	size_t _outputSize
-) noexcept
+void call(evm_result* o_result, evm_env* _opaqueEnv, evm_message const* _msg) noexcept
```
Comment (gumb0) [inline libevm/JitVM.cpp]: This function is now over 70 lines long, and doing two very distinct tasks, performing create and call. Consider splitting it into at least 2 internal functions

File: libevm/JitVM.cpp (hunk 2 above)
Comment (gumb0) [inline libevm/JitVM.cpp]: Yeah, there's one allocation less and one indirection less, but in general such esoteric tricks should better be used only when we're really sure we need them. 

In general placement new should better not be used outside of lower level memory-managing utility classes like containers, I guess.

But hm, ok if you wanted to try it  :) I'm ok with leaving it like this for now

**Comment by chfast** [comment]
Requested changes applied :).

File: libevm/JitVM.cpp [hunk 5]
```diff
@@ -127,61 +127,90 @@ void getTxContext(evm_tx_context* result, evm_env* _opaqueEnv) noexcept
...
 	*o_hash = toEvmC(env.blockHash(_number));
 }
 
-int64_t call(
-	evm_env* _opaqueEnv,
-	evm_message const* _msg,
-	uint8_t* _outputData,
-	size_t _outputSize
-) noexcept
+void create(evm_result* o_result, ExtVMFace& env, evm_message const* _msg) noexcept
```
Comment (gumb0) [inline libevm/JitVM.cpp]: _env

**Pull request merged**